File: 200/207 - Course Schedule.java
import java.util.ArrayList;

public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> prereq;
        List<Integer> curr;
        int[] isVisited;
        int i;
        
        if (numCourses == 0)
            return true;
        System.out.println(numCourses);
        prereq = new ArrayList<List<Integer>>(numCourses);
        isVisited = new int[numCourses];
        
        for (i = 0; i < numCourses; ++i)
            prereq.add(null);
        
        for (int[] pair : prerequisites) {
            if ((curr = prereq.get(pair[0])) == null) {
                curr = new ArrayList<Integer>();
                prereq.set(pair[0], curr);
            }
            curr.add(pair[1]);
        }
        
        for (i = 0; i < numCourses; ++i) {
            if (! depthFirstSearch(prereq, isVisited, i))
                return false;
        }
        
        return true;
    }
    
    private boolean depthFirstSearch(List<List<Integer>> prereq, int[] isVisited, int index) {
        if (isVisited[index] == 2 || prereq.get(index) == null)
            return true;
        if (isVisited[index] == 1)
            return false;
        isVisited[index] = 1;
        for (int p : prereq.get(index)) {
            if (! depthFirstSearch(prereq, isVisited, p))
                return false;
        }
        isVisited[index] = 2;
        return true;
    }
}

File: csharp/207.course-schedule.cs
/*
 * @lc app=leetcode id=207 lang=csharp
 *
 * [207] Course Schedule
 */

using System.Collections.Generic;
using System.Linq;

public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites) {
        var deps = new IList<int>[numCourses];
        var count = new int[numCourses];
        var queue = new Queue<int>();
        foreach (int[] pair in prerequisites) {
            int f = pair[0], s = pair[1];
            IList<int> dep = deps[s];
            if (dep == null) {
                dep = new List<int>();
                deps[s] = dep;
            }
            dep.Add(f);
            ++count[f];
        }
        for (int i = 0; i < numCourses; ++i) {
            if (count[i] == 0) {
                queue.Enqueue(i);
            }
        }
        while (queue.Count > 0) {
            int curr = queue.Dequeue();
            if (deps[curr] != null) {
                foreach (int next in deps[curr]) {
                    --count[next];
                    if (count[next] == 0) {
                        queue.Enqueue(next);
                    }
                }
            }
        }
        return count.All(x => x == 0);
    }
}

